### PR TITLE
APPS/IO-DEMO: Improve logging

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Improve logging by printing the type of operation that was completed/canceled.

## Why ?

To improve logging in IO-demo.

## How ?

1. Define global constants which have names of the UCP operations.
2. Save the name of the operation to the request structure - `struct ucx_request`/
3. Print the name of the operation when canceling operations or completing them during disconnect.

### CODE REVIEW

**Review by yosefe — COMMENTED** [review]
why cancel_all does not make receives complete immediately?

**Comment by dmitrygx** [comment]
> why cancel_all does not make receives complete immediately?

@yosefe it seems that those request were already matched (e.g. RNDV GET Zcopy protocol), but not completed.
```
[1623660842.372670] [jazz29:26581:0]     ib_mlx5_log.c:147  UCX  DIAG  Transport retry count exceeded on mlx5_3:1/RoCE (synd 0x15 vend 0x81 hw_synd 0/0)
[1623660842.372670] [jazz29:26581:0]     ib_mlx5_log.c:147  UCX  DIAG  DCI QP 0x9dc wqe[254]: SEND s-e [rqpn 0x1c3e rmac ec:0d:9a:8c:cd:4f sgix 3 dgid ::ffff:2.1.3.28 tc 106] [inl len 18] [va 0x7f86dde63702 len 8238 lkey 0x2e4138]
[1623660842.372980] [UCX-connection #31 2.1.3.28:20002] detected error: Endpoint timeout
[1623660842.372998] [UCX] removed [UCX-connection #31 2.1.3.28:20002] from connection map
[1623660842.373009] [DEMO] disconnecting connection 0x1585ab0 with 16 uncompleted operations (read: 3054/3060; write: 3192/3202) due to "Endpoint timeout"
[1623660842.373016] [UCX-connection #31 2.1.3.28:20002] destroying, ep is 0x7f86f483c1d0
[1623660842.373020] [UCX-connection #31 2.1.3.28:20002] closing ep 0x7f86f483c1d0 mode force
[1623660842.373038] [UCX-connection #31 2.1.3.28:20002] canceling 0x1776590 (ucp_tag_send_nb) request #1
[1623660842.373044] [UCX-connection #31 2.1.3.28:20002] canceling 0x16c4710 (ucp_tag_send_nb) request #2
[1623660842.373047] [UCX-connection #31 2.1.3.28:20002] canceling 0x16ff010 (ucp_tag_recv_nb) request #3
[1623660842.373053] [UCX-connection #31 2.1.3.28:20002] canceling 0x1621810 (ucp_tag_recv_nb) request #4
[1623660842.373056] [UCX-connection #31 2.1.3.28:20002] canceling 0x16c0e10 (ucp_tag_recv_nb) request #5
[1623660842.373062] [UCX-connection #31 2.1.3.28:20002] completing request 0x16c0e10 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373066] [UCX-connection #31 2.1.3.28:20002] canceling 0x1700690 (ucp_tag_recv_nb) request #6
[1623660842.373072] [UCX-connection #31 2.1.3.28:20002] completing request 0x1700690 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373076] [UCX-connection #31 2.1.3.28:20002] canceling 0x16caa10 (ucp_tag_recv_nb) request #7
[1623660842.373081] [UCX-connection #31 2.1.3.28:20002] completing request 0x16caa10 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373084] [UCX-connection #31 2.1.3.28:20002] canceling 0x171c310 (ucp_tag_recv_nb) request #8
[1623660842.373087] [UCX-connection #31 2.1.3.28:20002] completing request 0x171c310 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373090] [UCX-connection #31 2.1.3.28:20002] canceling 0x16bfd90 (ucp_tag_recv_nb) request #9
[1623660842.373094] [UCX-connection #31 2.1.3.28:20002] completing request 0x16bfd90 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373097] [UCX-connection #31 2.1.3.28:20002] canceling 0x1779290 (ucp_tag_recv_nb) request #10
[1623660842.373101] [UCX-connection #31 2.1.3.28:20002] completing request 0x1779290 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373104] [UCX-connection #31 2.1.3.28:20002] canceling 0x161ad90 (ucp_tag_recv_nb) request #11
[1623660842.373110] [UCX-connection #31 2.1.3.28:20002] completing request 0x161ad90 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373113] [UCX-connection #31 2.1.3.28:20002] canceling 0x16a4890 (ucp_tag_recv_nb) request #12
[1623660842.373117] [UCX-connection #31 2.1.3.28:20002] completing request 0x16a4890 (ucp_tag_recv_nb) with status "Request canceled" (-16) during disconnect
[1623660842.373120] [UCX-connection #31 2.1.3.28:20002] completing 0x16ff010 (ucp_tag_recv_nb) manually with "Endpoint timeout"
[1623660842.373126] [UCX-connection #31 2.1.3.28:20002] completing request 0x16ff010 (ucp_tag_recv_nb) with status "Endpoint timeout" (-80) during disconnect
[1623660842.373133] [UCX-connection #31 2.1.3.28:20002] completing 0x1621810 (ucp_tag_recv_nb) manually with "Endpoint timeout"
[1623660842.373139] [UCX-connection #31 2.1.3.28:20002] completing request 0x1621810 (ucp_tag_recv_nb) with status "Endpoint timeout" (-80) during disconnect
```

**Comment by yosefe** [comment]
> @yosefe it seems that those request were already matched (e.g. RNDV GET Zcopy protocol), but not completed.
> Here in the example 2/12 requests weren't canceled, so we completed them manually.

if already matched, I would expect that moving QP to error state (from ep_close(UCP_EP_CLOSE_MODE_FORCE)) would make RDMA_READ complete with error

**Comment by dmitrygx** [comment]
> > @yosefe it seems that those request were already matched (e.g. RNDV GET Zcopy protocol), but not completed.
> > Here in the example 2/12 requests weren't canceled, so we completed them manually.
> 
> if already matched, I would expect that moving QP to error state (from ep_close(UCP_EP_CLOSE_MODE_FORCE)) would make RDMA_READ complete with error

I see, it should then.
I'm going to check what kind of requests are not completed.
BTW, multi AM Bcopy/Zcopy can be problematic, they can send a few fragments, so the request will be matched on the receiver, but the remaining fragments won't be sent due to error.

**Comment by yosefe** [comment]
> > > @yosefe it seems that those request were already matched (e.g. RNDV GET Zcopy protocol), but not completed.
> > > Here in the example 2/12 requests weren't canceled, so we completed them manually.
> > 
> > 
> > if already matched, I would expect that moving QP to error state (from ep_close(UCP_EP_CLOSE_MODE_FORCE)) would make RDMA_READ complete with error
> 
> I see, it should then.
> I'm going to check what kind of requests are not completed.
> BTW, multi AM Bcopy/Zcopy can be problematic, they can send a few fragments, so the request will be matched on the receiver, but the remaining fragments won't be sent due to error.

Currently they are disabled because of this reason; with TAG, will add ep_ptr to int4 to cancel them; and with UCP AM we could also cancel the receives and ignore additional fragments by ep_ptr

**Comment by dmitrygx** [comment]
replaced by #6953 - the fix should be done in UCP code.
@yosefe wdyt to add the logging from the PR in IO-demo? it helps me to debug the issue and understand the type of requests and what happens with them.

**Comment by dmitrygx** [comment]
@brminich @hoopoepg @yosefe could you review pls?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 1]
```diff
@@ -1057,8 +1068,9 @@ bool UcxConnection::process_request(const char *what,
             return status == UCS_OK;
         } else {
             // will be completed by callback
-            r->callback = callback;
-            r->conn     = this;
+            r->callback     = callback;
+            r->conn         = this;
+            r->what         = what;
```
Comment (hoopoepg) [inline test/apps/iodemo/ucx_wrapper.cc:1073]: really strange alignment

Comment (dmitrygx) [reply test/apps/iodemo/ucx_wrapper.cc]: thanks, fixed

**Review by hoopoepg — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Review by brminich — APPROVED** [review]

File: test/apps/iodemo/ucx_wrapper.cc [hunk 2]
```diff
@@ -19,6 +19,13 @@
 #define AM_MSG_ID 0
 
 
+const char *tag_recv_nb_name      = "ucp_tag_recv_nb";
+const char *tag_send_nb_name      = "ucp_tag_send_nb";
+const char *am_send_nbx_name      = "ucp_am_send_nbx";
+const char *am_recv_data_nbx_name = "ucp_am_recv_data_nbx";
+const char *conn_id_recv_name     = "conn_id receive";
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:26]: why need to add these vars? each of them is used only once

File: test/apps/iodemo/ucx_wrapper.cc [hunk 3]
```diff
@@ -987,9 +998,9 @@ void UcxConnection::request_completed(ucx_request *r)
     ucs_list_del(&r->pos);
 
     if (_disconnect_cb != NULL) {
-        UCX_CONN_LOG << "completing request " << r << " with status \""
-                     << ucs_status_string(r->status) << "\" (" << r->status
-                     << ")" << " during disconnect";
+        UCX_CONN_LOG << "completing request " << r << " (" << r->what
+                     << ") with status \"" << ucs_status_string(r->status)
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:1002]: maybe it should be like this:
"completing ucp_tag_recv_nb request 0x12345 with status 'Connection reset' during disconnect"

File: test/apps/iodemo/ucx_wrapper.cc (hunk 2 above)
Comment (dmitrygx) [reply test/apps/iodemo/ucx_wrapper.cc:26]: to save its address in request's field

Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc:26]: we can save the pointer to the string literal itself

**Review by yosefe — APPROVED** [review]

File: test/apps/iodemo/ucx_wrapper.cc (hunk 2 above)
Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc]: need to treat these as `const char *`

**Pull request merged**